File: src/main/java/dark/leech/text/plugin/vbook/PluginZipExtractor.java
package dark.leech.text.plugin.vbook;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;

import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.model.FileHeader;

import dark.leech.text.action.Log;
import dark.leech.text.plugin.security.NetworkSecurityValidator;
import dark.leech.text.plugin.security.RegexSecurityValidator;
import dark.leech.text.plugin.security.SecurityValidationException;
import dark.leech.text.plugin.security.ZipSecurityValidator;
import dark.leech.text.plugin.vbook.exception.VBookPluginException;
import dark.leech.text.plugin.vbook.model.VBookPluginEntity;
import dark.leech.text.util.FileUtils;

/**
 * Extracts and parses vBook plugin.zip files. Downloads ZIP from URL and extracts plugin.json,
 * scripts, and icon.
 */
public class PluginZipExtractor {

    private static final String DEFAULT_CACHE_DIR = "tools/plugins/vbook-cache";
    private static final Gson GSON = new Gson();

    /**
     * Extract plugin from local ZIP file.
     *
     * @param zipPath Path to plugin.zip file
     * @return VBookPluginEntity with extracted data
     * @throws VBookPluginException if extraction fails
     */
    public VBookPluginEntity extractFromZip(Path zipPath) throws VBookPluginException {
        File zipFile = zipPath.toFile();
        if (!zipFile.exists()) {
            throw new VBookPluginException("ZIP file not found: " + zipPath);
        }

        // Validate ZIP file for security threats (after existence check)
        try {
            ZipSecurityValidator.validateZipFile(zipPath);
        } catch (SecurityValidationException e) {
            throw new VBookPluginException("ZIP security validation failed: " + e.getMessage(), e);
        }

        try {
            return extractZipInternal(zipFile);
        } catch (VBookPluginException e) {
            // Re-throw VBookPluginException as-is
            throw e;
        } catch (Exception e) {
            throw new VBookPluginException("Failed to extract plugin from: " + zipPath, e);
        }
    }

    /**
     * Download and extract plugin from URL.
     *
     * @param zipUrl URL to plugin.zip file
     * @return VBookPluginEntity with extracted data
     * @throws VBookPluginException if download or extraction fails
     */
    public VBookPluginEntity extractFromZip(URL zipUrl) throws VBookPluginException {
        try {
            // Download to cache
            Path cacheDir = Paths.get(FileUtils.validate(DEFAULT_CACHE_DIR));
            Files.createDirectories(cacheDir);

            String filename = extractFilenameFromUrl(zipUrl.toString());
            Path localZipPath = cacheDir.resolve(filename);

            // Remove existing file if exists (forces fresh download)
            try {
                Files.deleteIfExists(localZipPath);
            } catch (IOException ignored) {
            }

            downloadFile(zipUrl, localZipPath);
            return extractFromZip(localZipPath);

        } catch (Exception e) {
            throw new VBookPluginException("Failed to download plugin from: " + zipUrl, e);
        }
    }

    /**
     * Internal extraction logic using zip4j. Includes ZIP Slip protection to prevent path traversal
     * attacks.
     */
    private VBookPluginEntity extractZipInternal(File zipFile) throws Exception {
        ZipFile zip = new ZipFile(zipFile);

        // Validate ZIP entries for path traversal (ZIP Slip protection)
        validateZipEntries(zip);

        // Extract plugin.json (metadata)
        String pluginJson = extractEntryAsString(zip, "plugin.json");
        if (pluginJson == null) {
            throw new VBookPluginException("plugin.json not found in ZIP");
        }

        // Parse plugin.json and create entity
        VBookPluginEntity entity = parsePluginMetadata(pluginJson);

        // Validate regex pattern from metadata
        if (entity.getRegexp() != null) {
            validateRegexPattern(entity.getRegexp());
        }

        // Extract scripts from src/ directory
        Map<String, String> scriptContents = extractScripts(zip);
        entity.setScriptContents(scriptContents);

        // Extract icon.png
        String iconBase64 = extractIconAsBase64(zip);
        if (iconBase64 != null) {
            entity.setIconBase64(iconBase64);
        }

        // Set raw metadata
        entity.setRawMetadata(pluginJson);

        // Validate required scripts
        if (!scriptContents.containsKey("chap") && !scriptContents.containsKey("toc") && !scriptContents.containsKey("gen")) {
            throw new VBookPluginException("Plugin must have at least chap.js or toc.js");
        }

        return entity;
    }

    /**
     * Validate ZIP entries to prevent ZIP Slip path traversal attacks. Uses ZipSecurityValidator
     * for comprehensive security checks.
     */
    private void validateZipEntries(ZipFile zip) throws VBookPluginException {
        try {
            for (net.lingala.zip4j.model.FileHeader header : zip.getFileHeaders()) {
                String fileName = header.getFileName();
                ZipSecurityValidator.validateEntryName(fileName);
            }
        } catch (SecurityValidationException e) {
            throw new VBookPluginException("ZIP entry validation failed", e);
        } catch (Exception e) {
            throw new VBookPluginException("Failed to validate ZIP entries", e);
        }
    }

    /**
     * Validate regex pattern to prevent ReDoS attacks. Uses RegexSecurityValidator for
     * comprehensive timeout and pattern checks.
     */
    private void validateRegexPattern(String regex) throws VBookPluginException {
        try {
            RegexSecurityValidator.validateRegex(regex);
        } catch (SecurityValidationException e) {
            throw new VBookPluginException("Regex validation failed", e);
        }
    }

    /** Parse plugin.json metadata into entity. */
    private VBookPluginEntity parsePluginMetadata(String json) {
        try {
            JsonObject root = GSON.fromJson(json, JsonObject.class);

            // Create entity with default values
            VBookPluginEntity entity = new VBookPluginEntity();
            entity.setScripts(new HashMap<>());
            entity.setScriptContents(new HashMap<>());

            // Parse metadata section
            JsonObject metadata = root.getAsJsonObject("metadata");
            if (metadata != null) {
                entity.setName(getString(metadata, "name"));
                entity.setAuthor(getString(metadata, "author"));
                entity.setVersion(getInt(metadata, "version"));
                entity.setSource(getString(metadata, "source"));
                entity.setRegexp(getString(metadata, "regexp"));
                entity.setDescription(getString(metadata, "description"));
                entity.setLocale(getString(metadata, "locale"));
                entity.setType(getString(metadata, "type"));
                entity.setLanguage(getString(metadata, "language"));
                entity.setPriority(getInt(metadata, "priority"));
                entity.setTag(getString(metadata, "tag"));
            }

            // Parse script section (file references)
            JsonObject scripts = root.getAsJsonObject("script");
            if (scripts != null) {
                Map<String, String> scriptMap = new HashMap<>();
                scripts.entrySet()
                        .forEach(
                                entry -> {
                                    scriptMap.put(entry.getKey(), entry.getValue().getAsString());
                                });
                entity.setScripts(scriptMap);
            }

            // Parse config section if present
            JsonObject config = root.getAsJsonObject("config");
            if (config != null) {
                // Config options can be stored in description for now
                String configStr = config.toString();
                String currentDesc = entity.getDescription() != null ? entity.getDescription() : "";
                entity.setDescription(currentDesc + "\n\nConfig: " + configStr);
            }

            return entity;

        } catch (JsonSyntaxException e) {
            throw new VBookPluginException("Failed to parse plugin.json", e);
        }
    }

    /** Extract all .js files from src/ directory. */
    private Map<String, String> extractScripts(ZipFile zip) throws Exception {
        Map<String, String> scripts = new HashMap<>();
        Pattern scriptPattern = Pattern.compile("^src/([^/]+)\\.js$");

        List<FileHeader> headers = zip.getFileHeaders();
        for (FileHeader header : headers) {
            String entryName = header.getFileName();

            // Skip directories
            if (header.isDirectory()) {
                continue;
            }

            // Match src/*.js files
            Matcher matcher = scriptPattern.matcher(entryName);
            if (matcher.matches()) {
                String scriptName = matcher.group(1);
                String content = extractEntryAsString(zip, entryName);
                if (content != null) {
                    scripts.put(scriptName, content);
                }
            }
        }

        return scripts;
    }

    /** Extract icon.png and convert to base64 data URI. */
    private String extractIconAsBase64(ZipFile zip) {
        try {
            List<FileHeader> headers = zip.getFileHeaders();
            for (FileHeader header : headers) {
                if ("icon.png".equals(header.getFileName()) && !header.isDirectory()) {
                    try (InputStream is = zip.getInputStream(header)) {
                        byte[] bytes = is.readAllBytes();
                        String base64 = Base64.getEncoder().encodeToString(bytes);
                        return "data:image/png;base64," + base64;
                    }
                }
            }
        } catch (Exception e) {
            // Icon extraction is optional - log but don't fail
            Log.add("Failed to extract icon (non-critical): " + e.getMessage());
        }
        return null;
    }

    /** Extract a single entry from ZIP as string. */
    private String extractEntryAsString(ZipFile zip, String entryName) throws Exception {
        FileHeader header = zip.getFileHeader(entryName);
        if (header == null || header.isDirectory()) {
            return null;
        }

        InputStream is = zip.getInputStream(header);
        return new String(is.readAllBytes(), StandardCharsets.UTF_8);
    }

    /** Download file from URL to local path with security validation. */
    private void downloadFile(URL url, Path destPath) throws IOException {
        // Validate URL before download
        try {
            NetworkSecurityValidator.validateUrl(url.toString());
        } catch (SecurityValidationException e) {
            throw new IOException("URL validation failed: " + e.getMessage(), e);
        }

        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setConnectTimeout(30000);
        connection.setReadTimeout(30000);
        connection.setInstanceFollowRedirects(true);

        // Validate response headers
        String contentType = connection.getContentType();
        long contentLength = connection.getContentLengthLong();

        try {
            NetworkSecurityValidator.validateResponse(contentType, contentLength);
        } catch (SecurityValidationException e) {
            connection.disconnect();
            throw new IOException("Response validation failed: " + e.getMessage(), e);
        }

        // Download to temporary file first
        Path tempPath = destPath.getParent().resolve(destPath.getFileName().toString() + ".tmp");
        try {
            try (InputStream is = connection.getInputStream()) {
                Files.copy(is, tempPath, StandardCopyOption.REPLACE_EXISTING);
            }

            // Atomic move from temp to final destination
            Files.move(
                    tempPath,
                    destPath,
                    StandardCopyOption.ATOMIC_MOVE,
                    StandardCopyOption.REPLACE_EXISTING);

        } catch (Exception e) {
            // Clean up temp file on failure
            try {
                Files.deleteIfExists(tempPath);
            } catch (IOException ignored) {
                // Ignore cleanup failures
            }
            throw e;
        } finally {
            connection.disconnect();
        }
    }

    /** Extract filename from URL. */
    private String extractFilenameFromUrl(String url) {
        String[] parts = url.split("/");
        return parts[parts.length - 1];
    }

    // Helper methods for JSON parsing

    private String getString(JsonObject obj, String name) {
        if (obj.has(name) && !obj.get(name).isJsonNull()) {
            return obj.get(name).getAsString();
        }
        return null;
    }

    private Integer getInt(JsonObject obj, String name) {
        if (obj.has(name) && !obj.get(name).isJsonNull()) {
            return obj.get(name).getAsInt();
        }
        return null;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/vbook/PluginZipExtractor.kt
package dev.haipham22.leechtext.plugin.vbook

import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.plugin.js.api.HttpSpec
import dev.haipham22.leechtext.plugin.js.api.httpExecute
import dev.haipham22.leechtext.plugin.security.NetworkSecurityValidator
import dev.haipham22.leechtext.plugin.security.RegexSecurityValidator
import dev.haipham22.leechtext.plugin.security.SecurityValidationException
import dev.haipham22.leechtext.plugin.security.ZipSecurityValidator
import dev.haipham22.leechtext.plugin.vbook.exception.VBookPluginException
import dev.haipham22.leechtext.plugin.vbook.model.VBookPluginEntity
import dev.haipham22.leechtext.util.EnginePaths
import dev.haipham22.leechtext.util.ZipIo
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okio.FileSystem
import okio.Path
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/**
 * Extract + parse vBook plugin.zip (port từ PluginZipExtractor.java; P5.2c common —
 * zip qua ZipIo seam, download qua HttpEngine). ZIP Slip + ReDoS protection qua
 * security validators.
 */
class PluginZipExtractor(
    private val log: EngineLogger,
) {
    private val fs: FileSystem get() = FileSystem.SYSTEM

    /** Extract plugin từ local ZIP file. */
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    fun extractFromZip(zipPath: Path): VBookPluginEntity {
        if (!fs.exists(zipPath)) {
            throw VBookPluginException("ZIP file not found: $zipPath")
        }

        // Validate ZIP security threats (sau existence check)
        try {
            ZipSecurityValidator.validateZipFile(zipPath, log)
        } catch (e: SecurityValidationException) {
            throw VBookPluginException("ZIP security validation failed: ${e.message}", e)
        }

        return try {
            extractZipInternal(zipPath)
        } catch (e: VBookPluginException) {
            throw e
        } catch (e: Exception) {
            throw VBookPluginException("Failed to extract plugin from: $zipPath", e)
        }
    }

    /** Download + extract plugin từ URL. */
    fun extractFromZip(zipUrl: String): VBookPluginEntity = try {
        // Download vào cache
        val cacheDir = EnginePaths.dataDir / "tools" / "plugins" / "vbook-cache"
        fs.createDirectories(cacheDir)

        val filename = extractFilenameFromUrl(zipUrl)
        val localZipPath = cacheDir / filename

        // Remove file cũ nếu có (forces fresh download)
        runCatching { if (fs.exists(localZipPath)) fs.delete(localZipPath) }

        downloadFile(zipUrl, localZipPath)
        log.debug("[extractFromZip] downloaded ${fs.metadataOrNull(localZipPath)?.size ?: 0} bytes → $localZipPath")
        extractFromZip(localZipPath)
    } catch (e: Exception) {
        throw VBookPluginException("Failed to download plugin from: $zipUrl", e)
    }

    /** Extraction logic nội bộ (ZipIo) — có ZIP Slip protection. */
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    private fun extractZipInternal(zipFile: Path): VBookPluginEntity {
        // Validate entries chống path traversal (ZIP Slip)
        validateZipEntries(zipFile)

        // Extract plugin.json (metadata)
        val pluginJson =
            extractEntryAsString(zipFile, "plugin.json")
                ?: throw VBookPluginException("plugin.json not found in ZIP")

        // Parse plugin.json → entity
        val entity = parsePluginMetadata(pluginJson)

        // Plugin mã hóa (repo bên thứ 3 flag metadata "encrypt": true) — script là
        // ciphertext, execute làm JS chỉ ra ReferenceError rác. Chặn tại install.
        val encryptFlag =
            runCatching {
                Json.parseToJsonElement(pluginJson).jsonObject["metadata"]
                    ?.jsonObject?.get("encrypt")?.jsonPrimitive?.booleanOrNull
            }.getOrNull()
        if (encryptFlag == true) {
            throw VBookPluginException("Encrypted plugin (encrypt: true) — decrypt not supported", code = "encrypted")
        }

        // Validate regex pattern từ metadata
        entity.regexp?.let { validateRegexPattern(it) }

        // Extract scripts từ src/
        val scriptContents = extractScripts(zipFile)
        entity.scriptContents = scriptContents

        // Extract icon.png
        extractIconAsBase64(zipFile)?.let { entity.iconBase64 = it }

        // Set raw metadata
        entity.rawMetadata = pluginJson

        // Validate required scripts
        if (!scriptContents.containsKey("chap") &&
            !scriptContents.containsKey("toc") &&
            !scriptContents.containsKey("gen")
        ) {
            throw VBookPluginException("Plugin must have at least chap.js or toc.js")
        }

        return entity
    }

    /** Validate ZIP entries chống ZIP Slip. */
    private fun validateZipEntries(zip: Path) {
        try {
            for (entry in ZipIo.entries(zip, log)) {
                ZipSecurityValidator.validateEntryName(entry.name)
            }
        } catch (e: SecurityValidationException) {
            throw VBookPluginException("ZIP entry validation failed", e)
        } catch (e: Exception) {
            throw VBookPluginException("Failed to validate ZIP entries", e)
        }
    }

    /** Validate regex pattern chống ReDoS. */
    private fun validateRegexPattern(regex: String) {
        try {
            RegexSecurityValidator.validateRegex(regex)
        } catch (e: SecurityValidationException) {
            throw VBookPluginException("Regex validation failed", e)
        }
    }

    /** Parse plugin.json metadata → entity. */
    private fun parsePluginMetadata(json: String): VBookPluginEntity = try {
        val root = Json.parseToJsonElement(json).jsonObject
        val entity =
            VBookPluginEntity(
                scripts = HashMap(),
                scriptContents = HashMap(),
            )

        // Metadata section
        (root["metadata"] as? JsonObject)?.let { metadata ->
            entity.name = getString(metadata, "name")
            entity.author = getString(metadata, "author")
            entity.version = getInt(metadata, "version")
            entity.source = getString(metadata, "source")
            entity.regexp = getString(metadata, "regexp")
            entity.description = getString(metadata, "description")
            entity.locale = getString(metadata, "locale")
            entity.type = getString(metadata, "type")
            entity.language = getString(metadata, "language")
            entity.priority = getInt(metadata, "priority")
            entity.tag = getString(metadata, "tag")
        }

        // Script section (file references)
        (root["script"] as? JsonObject)?.let { scripts ->
            val scriptMap = HashMap<String, String>()
            for ((key, value) in scripts) {
                scriptMap[key] = value.jsonPrimitive.content
            }
            entity.scripts = scriptMap
        }

        // Config section (vBook extension-api) — key → spec cho UI + JS injection
        (root["config"] as? JsonObject)?.let { config ->
            for ((key, specJson) in config) {
                val spec = specJson as? JsonObject ?: continue
                entity.config[key] =
                    dev.haipham22.leechtext.plugin.vbook.model.VBookConfigSpec(
                        title = getString(spec, "title"),
                        subtitle = getString(spec, "subtitle"),
                        default = getString(spec, "default"),
                        values = (spec["values"] as? JsonArray)?.map { it.jsonPrimitive.content } ?: emptyList(),
                        mode = getString(spec, "mode"),
                        format = getString(spec, "format"),
                    )
            }
        }

        entity
    } catch (e: SerializationException) {
        throw VBookPluginException("Failed to parse plugin.json", e)
    }

    /** Extract mọi file .js trong src/. */
    private fun extractScripts(zip: Path): MutableMap<String, String> {
        val scripts = HashMap<String, String>()
        val scriptPattern = Regex("^src/([^/]+)\\.js$")

        for (entry in ZipIo.entries(zip, log)) {
            // Skip directories
            if (entry.isDirectory) continue

            // Match src/*.js
            val matcher = scriptPattern.matchEntire(entry.name) ?: continue
            val scriptName = matcher.groupValues[1]
            val content = extractEntryAsString(zip, entry.name)
            if (content != null) {
                scripts[scriptName] = content
            }
        }

        return scripts
    }

    /** Extract icon.png → base64 data URI. */
    @OptIn(ExperimentalEncodingApi::class)
    private fun extractIconAsBase64(zip: Path): String? {
        return try {
            if (ZipIo.entries(zip, log).any { it.name == "icon.png" && !it.isDirectory }) {
                val bytes = ZipIo.readEntry(zip, "icon.png", log) ?: return null
                val base64 = Base64.encode(bytes)
                "data:image/png;base64,$base64"
            } else {
                null
            }
        } catch (e: Exception) {
            // Icon optional — log nhưng không fail
            log.add("Failed to extract icon (non-critical): ${e.message}")
            null
        }
    }

    /** Extract một entry từ ZIP thành string. */
    private fun extractEntryAsString(
        zip: Path,
        entryName: String,
    ): String? = ZipIo.readEntry(zip, entryName, log)?.decodeToString()

    /** Download file từ URL xuống local path với security validation. */
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    private fun downloadFile(
        url: String,
        destPath: Path,
    ) {
        // Validate URL trước khi download
        try {
            NetworkSecurityValidator.validateUrl(url)
        } catch (e: SecurityValidationException) {
            throw VBookPluginException("URL validation failed: ${e.message}", e)
        }

        val result =
            try {
                httpExecute(HttpSpec(method = "GET", url = url), log)
            } catch (e: Exception) {
                throw VBookPluginException("Download failed: ${url.take(120)}", e)
            }

        // Validate response headers
        try {
            NetworkSecurityValidator.validateResponse(result.headers["Content-Type"], result.body.size.toLong())
        } catch (e: SecurityValidationException) {
            throw VBookPluginException("Response validation failed: ${e.message}", e)
        }

        // Download vào temp file trước, atomic move từ temp → dest
        val tempPath = destPath.parent?.resolve(destPath.name + ".tmp") ?: destPath
        try {
            tempPath.parent?.let { fs.createDirectories(it) }
            fs.write(tempPath) { write(result.body) }
            fs.atomicMove(tempPath, destPath)
        } catch (e: Exception) {
            // Cleanup temp file khi fail
            runCatching { if (fs.exists(tempPath)) fs.delete(tempPath) }
            throw e
        }
    }

    /** Filename từ URL. */
    private fun extractFilenameFromUrl(url: String): String = url.split("/").last()
}

// Helper JSON parsing

private fun getString(
    obj: JsonObject,
    name: String,
): String? = (obj[name] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content

private fun getInt(
    obj: JsonObject,
    name: String,
): Int? = (obj[name] as? JsonPrimitive)?.takeIf { it !is JsonNull }?.content?.toIntOrNull()
